File: 2020/day/12/src/Main.hs
{-# LANGUAGE BangPatterns #-}
{-# LANGUAGE LambdaCase #-}
{-# OPTIONS_GHC -Wall -Wextra -Werror #-}

module Main where

import Data.Foldable (traverse_)
import Data.List (foldl') --, scanl')
import System.Environment (getArgs)
import Prelude hiding (Either (..))

main :: IO ()
main = (getArgs >>=) . traverse_ $ \filename -> do
  putStrLn filename

  instructions <- parseInstructions <$> readFile filename

  putStrLn $ "\tpart 1: manhattan distance from current location to end of instructions"
  putStrLn $ "\t" ++ show (part1 instructions)

  putStrLn $ "\tpart 2: manhattan distance from current location to end of waypoint instructions"
  putStrLn $ "\t" ++ show (part2 instructions)

parseInstructions :: String -> [Instruction]
parseInstructions = map parseInstruction . lines

parseInstruction :: String -> Instruction
parseInstruction (c : ns)
  | c == 'N' = Absolute (North d)
  | c == 'S' = Absolute (South d)
  | c == 'E' = Absolute (East d)
  | c == 'W' = Absolute (West d)
  | c == 'L' = Relative (Left t)
  | c == 'R' = Relative (Right t)
  | c == 'F' = Relative (Forward d)
  where
    d = read ns
    t = case d of
      90 -> Ninety
      180 -> OneEighty
      270 -> TwoSeventy
      _ -> error $ "illegal angle " ++ show d ++ ", expected turns in increments of 90 degrees"
parseInstruction line = error $ "illegal instruction " ++ show line

part1 :: [Instruction] -> Int
part1 = magnitude . absolute east

part2 :: [Instruction] -> Int
part2 = magnitude . waypoint (1, 10)

magnitude :: [Absolute] -> Int
magnitude = distance (0, 0) . foldl' move (0, 0)

distance :: Position -> Position -> Int
distance (lat, lon) (lat', lon') = abs (lat - lat') + abs (lon - lon')

move :: Position -> Absolute -> Position
move (!lat, !lon) = \case
  North dist -> (lat + dist, lon)
  South dist -> (lat - dist, lon)
  East dist -> (lat, lon + dist)
  West dist -> (lat, lon - dist)

absolute :: Heading -> [Instruction] -> [Absolute]
absolute dir = \case
  [] -> []
  Absolute inst : is -> inst : absolute dir is
  Relative (Left deg) : is -> absolute (left deg dir) is
  Relative (Right deg) : is -> absolute (right deg dir) is
  Relative (Forward dist) : is -> forward dir dist : absolute dir is
  where
    left :: Degrees -> Heading -> Heading
    left t (Compass t') = Compass $ invert t <> t'

    right :: Degrees -> Heading -> Heading
    right t (Compass t') = Compass $ t <> t'

    forward :: Heading -> Distance -> Absolute
    forward (Compass Zero) = North
    forward (Compass Ninety) = East
    forward (Compass OneEighty) = South
    forward (Compass TwoSeventy) = West

waypoint :: Position -> [Instruction] -> [Absolute]
waypoint pos = \case
  [] -> []
  Absolute inst : is -> waypoint (move pos inst) is
  Relative (Left deg) : is -> waypoint (rotate (invert deg) pos) is
  Relative (Right deg) : is -> waypoint (rotate deg pos) is
  Relative (Forward steps) : is -> North (fst pos * steps) : East (snd pos * steps) : waypoint pos is
  where
    rotate :: Degrees -> Position -> Position
    rotate Zero p = p
    rotate Ninety (lat, lon) = (- lon, lat)
    rotate OneEighty (lat, lon) = (- lat, - lon)
    rotate TwoSeventy (lat, lon) = (lon, - lat)

type Position = (Int, Int)

data Heading = Compass Degrees
  deriving (Show)

east :: Heading
east = Compass Ninety

data Instruction = Absolute Absolute | Relative Relative
  deriving (Show)

data Absolute
  = North Distance
  | South Distance
  | East Distance
  | West Distance
  deriving (Show)

data Relative
  = Left Degrees
  | Right Degrees
  | Forward Distance
  deriving (Show)

type Distance = Int

data Degrees = Zero | Ninety | OneEighty | TwoSeventy
  deriving (Show)

instance Semigroup Degrees where
  Zero <> t = t
  Ninety <> Ninety = OneEighty
  Ninety <> OneEighty = TwoSeventy
  Ninety <> TwoSeventy = Zero
  OneEighty <> OneEighty = Zero
  OneEighty <> TwoSeventy = Ninety
  TwoSeventy <> TwoSeventy = OneEighty
  t <> t' = t' <> t

instance Monoid Degrees where
  mempty = Zero

invert :: Degrees -> Degrees
invert Zero = Zero
invert Ninety = TwoSeventy
invert OneEighty = OneEighty
invert TwoSeventy = Ninety

File: 2020/day/12/src/main.rs
use std::error;

fn main() -> Result<(), Error> {
    for filename in std::env::args().skip(1) {
        println!("{}", filename);

        let instructions = parse_instructions(&std::fs::read_to_string(filename)?)?;

        println!("\tpart 1: manhattan distance from current location to end of instructions");
        println!("\t{}", part1(&instructions));

        println!(
            "\tpart 2: manhattan distance from current location to end of waypoint instructions"
        );
        println!("\t{}", part2(&instructions));
    }

    Ok(())
}

////////////////////////////////////////////////////////////////////////////////

fn parse_instructions(src: &str) -> Result<Vec<Instruction>, Error> {
    src.lines().map(parse_instruction).collect()
}

macro_rules! error {
    ($fmt:literal $(, $arg:expr)*) => {
        return Err(Error::from(format!($fmt $(, $arg)*)))
    }
}

fn parse_instruction(src: &str) -> Result<Instruction, Error> {
    let cnt = src[1..].parse()?;
    let deg = parse_degrees(cnt);

    Ok(match &src[0..1] {
        "N" => Move(cnt, 0),
        "E" => Move(0, cnt),
        "S" => Move(-cnt, 0),
        "W" => Move(0, -cnt),
        "L" => Turn(-deg?),
        "R" => Turn(deg?),
        "F" => Forward(cnt),
        action => error!(
            "Unexpected action {:?}, expected one of N/E/S/W/L/R/F",
            action
        ),
    })
}

fn parse_degrees(cnt: i32) -> Result<Degrees, Error> {
    Ok(match cnt {
        0 => Zero,
        90 => Ninety,
        180 => OneEighty,
        270 => TwoSeventy,
        _ => error!("Unexpected angle {}°, expected increments of 90°", cnt),
    })
}

////////////////////////////////////////////////////////////////////////////////

fn part1(instructions: &[Instruction]) -> i32 {
    let mut heading = Ninety;
    let mut latitude = 0;
    let mut longitude = 0;

    for &instruction in instructions {
        match instruction {
            Move(dlat, dlon) => {
                latitude += dlat;
                longitude += dlon;
            }
            Turn(deg) => heading += deg,
            Forward(delta) => match heading {
                Zero => latitude += delta,
                Ninety => longitude += delta,
                OneEighty => latitude -= delta,
                TwoSeventy => longitude -= delta,
            },
        }
    }

    latitude.abs() + longitude.abs()
}

////////////////////////////////////////////////////////////////////////////////

fn part2(instructions: &[Instruction]) -> i32 {
    let mut ship_lat = 0;
    let mut ship_lon = 0;
    let mut waypoint_lat = 1;
    let mut waypoint_lon = 10;

    for &instruction in instructions {
        match instruction {
            Move(dlat, dlon) => {
                waypoint_lat += dlat;
                waypoint_lon += dlon;
            }
            Turn(Zero) => {}
            Turn(Ninety) => {
                std::mem::swap(&mut waypoint_lat, &mut waypoint_lon);
                waypoint_lat *= -1;
            }
            Turn(OneEighty) => {
                waypoint_lat *= -1;
                waypoint_lon *= -1;
            }
            Turn(TwoSeventy) => {
                std::mem::swap(&mut waypoint_lat, &mut waypoint_lon);
                waypoint_lon *= -1;
            }
            Forward(delta) => {
                ship_lat += delta * waypoint_lat;
                ship_lon += delta * waypoint_lon;
            }
        }
    }

    ship_lat.abs() + ship_lon.abs()
}

////////////////////////////////////////////////////////////////////////////////

type Error = Box<dyn error::Error>;

#[derive(Clone, Copy, Debug)]
enum Instruction {
    Move(i32, i32),
    Turn(Degrees),
    Forward(i32),
}
use Instruction::*;

#[derive(Clone, Copy, Debug)]
enum Degrees {
    Zero,
    Ninety,
    OneEighty,
    TwoSeventy,
}
use Degrees::*;

impl std::ops::AddAssign for Degrees {
    fn add_assign(&mut self, other: Degrees) {
        *self = match (*self, other) {
            (Zero, deg) => deg,
            (deg, Zero) => deg,
            (Ninety, Ninety) => OneEighty,
            (Ninety, OneEighty) => TwoSeventy,
            (OneEighty, Ninety) => TwoSeventy,
            (Ninety, TwoSeventy) => Zero,
            (TwoSeventy, Ninety) => Zero,
            (OneEighty, OneEighty) => Zero,
            (OneEighty, TwoSeventy) => Ninety,
            (TwoSeventy, OneEighty) => Ninety,
            (TwoSeventy, TwoSeventy) => OneEighty,
        }
    }
}

impl std::ops::Neg for Degrees {
    type Output = Self;

    fn neg(self) -> Degrees {
        match self {
            Zero => Zero,
            Ninety => TwoSeventy,
            OneEighty => OneEighty,
            TwoSeventy => Ninety,
        }
    }
}
